File: src/benchmarks/microbenchmarks/mbench_1.py
#!/usr/bin/env python3
"""
*   @details:
*       1. Generate 64bit keys and 64bit values from an uniform random distribution to be inserted, removed or queried.
*       2. Evaluation of Insertion: from 0% -> 95% of the total capacity.
*       3. Evaluation of Queries: Positive and Negative questions (50/50).
*       4. Evaluation of Deletion: Remove a selections of keys until the usage reaches ~50%.
*
*   @author: Cristian Sandu <cristian.sandu@tum.de>
"""
import matplotlib.pyplot as plt
from collections import defaultdict

import subprocess, json, tempfile, os

class MBench_1:
    NAME_EXECUTABLE  = f'{os.path.dirname(os.path.abspath(__file__))}/mbench_1'
    RESULT_FILE_NAME = 'generic_output'
    PERF_RESULT      = 'perf_data.out'

    def __init__(self, config: dict):
        self.__config     = config
        self.__bresults   = []
    
    def __prepare_dir(self):
        os.makedirs(self.__outputpath, exist_ok=True)

    def __parse_results_performance(self, lines: list[str]) -> dict:
        data = defaultdict(lambda: [])
        for line in lines:
            function_name, thread_id, key_num, key, value, status, start_time, end_time = line.split(' ')
            data[function_name].append(int(int(end_time) - int(start_time)))
        return data

    def __parse_results_memory(self, lines: list[str]) -> dict:
        data = defaultdict(lambda: [])
        for line in lines:
            function_name, thread_id, key_num, memory_usage = line.split(' ')
            data[function_name].append(int(memory_usage))
        return data

    def __perform_bechmark(self, libso_path: str, config_path: str) -> list[str]:
        process = subprocess.Popen([MBench_1.NAME_EXECUTABLE, libso_path, config_path, MBench_1.RESULT_FILE_NAME])
        exit_code = process.wait()
        if exit_code != 0: raise Exception(f'MBench_1 filed during execution, exit-code: {exit_code}')
        
        return {
            'performance': self.__parse_results_performance(
                open(f'{MBench_1.RESULT_FILE_NAME}_performance.out', 'r').readlines()),
            'memory': self.__parse_results_memory(
                open(f'{MBench_1.RESULT_FILE_NAME}_memory.out', 'r').readlines())
        }

    def perform_benchmark(self, libso_path: str):
        self.__bresults = defaultdict(lambda: [])
        for threadnum in self.__config['threadnum']:
            with tempfile.NamedTemporaryFile(mode='w') as temp:
                # rescale for the given threads
                config = self.__config.copy()
                config['threadnum'] = threadnum
                json.dump(config, temp)
                temp.flush()
                for _ in range(self.__config['repetitions']):
                    self.__bresults[threadnum].append(self.__perform_bechmark(libso_path=libso_path, config_path=temp.name))

    def plot_results(self, result_name: str):
        self.__outputpath = f"{self.__config['output_path']}/{result_name}"
        self.__prepare_dir()

        self.__plot_dataset_performfill()
        self.__plot_dataset_performquery()
        self.__plot_dataset_performdeletion()

        self.__plot_memory_performfill()
    
    def __plot_memory_performfill(self):
        capacity       = self.__config['capacity']
        threadnum      = self.__config['threadnum']
        granularity    = self.__config['granularity']
        repetitions    = self.__config['repetitions']
        filling_factor = self.__config['performfill']['filling_factor']

        average_data = defaultdict(lambda: [0] * granularity)
        for threadnum, datasets in self.__bresults.items():
            for dataset in datasets:
                data = dataset['memory']['dataset_performfill']
                data_index, size_data = [], len(data)
                for i in range(0, granularity): 
                    if i == 0:  data_index.append(0)
                    else:       data_index.append(100*i/granularity)
                    left_range  = int(len(data) * i    /granularity)
                    right_range = int(len(data) * (i+1)/granularity)
                    average_data[threadnum][i] += sum(data[left_range : right_range]) / len(data[left_range : right_range])
            average_data[threadnum] = [x/granularity for x in average_data[threadnum]]
            average_data[threadnum] = [x/1024.0 for x in average_data[threadnum]]

        for threadnum in self.__bresults:
            plt.plot(data_index, average_data[threadnum], label = f'thread no. {threadnum}')
        plt.xlabel('percentage of ds. usage %')
        plt.ylabel('memory usage (kbytes)')
        plt.title(f'[mem_performfill] memory usage ds_insert (no.threads:{threadnum}: {repetitions}, gran.: {granularity}, cap.: {capacity}, fill fact.: {filling_factor})')
        plt.grid(True)
        plt.legend()

        plt.savefig(f'{self.__outputpath}/mem_performfill.png', dpi=300, bbox_inches = "tight")
        plt.show()
        plt.clf()

    def __plot_dataset_performfill(self):
        capacity       = self.__config['capacity']
        threadnum      = self.__config['threadnum']
        granularity    = self.__config['granularity']
        repetitions    = self.__config['repetitions']
        filling_factor = self.__config['performfill']['filling_factor']

        average_data = defaultdict(lambda: [0] * granularity)
        for threadnum, datasets in self.__bresults.items():
            for dataset in datasets:
                dataset = dataset['performance']

                data = dataset['dataset_performfill']
                data_index, size_data = [], len(data)
                for i in range(0, granularity): 
                    if i == 0:  data_index.append(0)
                    else:       data_index.append(100*i/granularity)
                    left_range  = int(len(data) * i    /granularity)
                    right_range = int(len(data) * (i+1)/granularity)
                    average_data[threadnum][i] += sum(data[left_range : right_range]) / len(data[left_range : right_range])
            average_data[threadnum] = [x/granularity for x in average_data[threadnum]]

        for threadnum in self.__bresults:
            plt.plot(data_index, average_data[threadnum], label = f'thread no. {threadnum}')
        plt.xlabel('percentage of ds. usage %')
        plt.ylabel('latency (ns)')
        plt.title(f'[dataset_performfill] latency ds_insert (no.threads:{threadnum}, rep.: {repetitions}, gran.: {granularity}, cap.: {capacity}, fill fact.: {filling_factor})')
        plt.grid(True)
        plt.legend()

        plt.savefig(f'{self.__outputpath}/performfill.png', dpi=300, bbox_inches = "tight")
        plt.show()
        plt.clf()

    def __plot_dataset_performquery(self):
        capacity       = self.__config['capacity']
        threadnum      = self.__config['threadnum']
        granularity    = self.__config['granularity']
        repetitions    = self.__config['repetitions']
        query_factor   = self.__config['performquery']['query_factor']
        success_factor = self.__config['performquery']['success_factor']

        average_data = defaultdict(lambda: [0] * granularity)
        for threadnum, datasets in self.__bresults.items():
            data_index, data_len = [], len(datasets[0]['performance']['dataset_performqueries'])
            for i in range(0, granularity): 
                if i == 0:  data_index.append(0)
                else:       data_index.append(data_len * i/granularity)

            for dataset in datasets:
                data  = dataset['performance']['dataset_performqueries']
                for i in range(0, granularity): 
                    left_range  = int(len(data) * i    /granularity)
                    right_range = int(len(data) * (i+1)/granularity)
                    if  len(data[left_range : right_range]) == 0: continue
                    average_data[threadnum][i] += sum(data[left_range : right_range]) / len(data[left_range : right_range])
            average_data[threadnum] = [x/granularity for x in average_data[threadnum]]

        for threadnum in self.__bresults:
            plt.plot(data_index, average_data[threadnum], label = f'thread no. {threadnum}')
        plt.xlabel('number of queries')
        plt.ylabel('latency (ns)')
        plt.title(f'[dataset_performquery] latency ds_read (no.threads{threadnum}, rep.: {repetitions}, gran.: {granularity}, cap.: {capacity}, rate: {query_factor}, succ.: {success_factor*100}%)')
        plt.grid(True)
        plt.legend()

        plt.savefig(f'{self.__outputpath}/performquery.png', dpi=300, bbox_inches = "tight")
        plt.show()
        plt.clf()

    def __plot_dataset_performdeletion(self):
        capacity        = self.__config['capacity']
        threadnum       = self.__config['threadnum']
        granularity     = self.__config['granularity']
        repetitions     = self.__config['repetitions']
        deletion_factor = self.__config['performdeletion']['deletion_factor']
        success_factor  = self.__config['performdeletion']['success_factor']

        average_data = defaultdict(lambda: [0] * granularity)
        for threadnum, datasets in self.__bresults.items():
            data_index, data_len = [], len(datasets[0]['performance']['dataset_performdeletion'])
            for i in range(0, granularity): 
                if i == 0:  data_index.append(0)
                else:       data_index.append(data_len * i/granularity)

            for dataset in datasets:
                dataset = dataset['performance']
                data = dataset['dataset_performdeletion']
                for i in range(0, granularity): 
                    left_range  = int(len(data) * i    /granularity)
                    right_range = int(len(data) * (i+1)/granularity)
                    average_data[threadnum][i] += sum(data[left_range : right_range]) / len(data[left_range : right_range])
            average_data[threadnum] = [x/granularity for x in average_data[threadnum]]

        for threadnum in self.__bresults:
            plt.plot(data_index, average_data[threadnum], label = f'thread no. {threadnum}')
        plt.xlabel('number of deletions')
        plt.ylabel('latency (ns)')
        plt.title(f'[dataset_performdeletion] latency ds_remove (no.threads:{threadnum}, rep.: {repetitions}, gran.: {granularity}, cap.: {capacity}, del fact.: {deletion_factor}, succ.: {success_factor})')
        plt.grid(True)
        plt.legend()

        plt.savefig(f'{self.__outputpath}/performdeletion.png', dpi=300, bbox_inches = "tight")
        plt.show()
        plt.clf()

    def __perform_perf_profilling(self, libso_path: str, config_path: str) -> dict:
        """
            Perform the perf_profilling over self.perform_benchmark. The execution
            is slow
        """
        process = subprocess.Popen(['perf', 'record', '-ag', '-e LLC-stores,LLC-store-misses,LLC-prefetch-misses', 
                                    MBench_1.NAME_EXECUTABLE, libso_path, config_path, MBench_1.RESULT_FILE_NAME])
        exit_code = process.wait()
        if exit_code != 0: raise Exception(f'MBench_1 filed during execution, exit-code: {exit_code}')

        process = subprocess.Popen(['perf', 'data', 'convert','--force', '--to-json', MBench_1.PERF_RESULT])
        exit_code = process.wait()
        if exit_code != 0: raise Exception(f'MBench_1 filed during execution, exit-code: {exit_code}')

        return json.load(open(MBench_1.PERF_RESULT, 'r'))
    
    def perform_perf_profilling(self, libso_path: str):
        with tempfile.NamedTemporaryFile(mode='w') as temp:
            json.dump(self.__config, temp)
            temp.flush()
            
            self.__bresults = []
            for _ in range(self.__config['perf']['repetitions']):
                result = self.__perform_perf_profilling(libso_path=libso_path, config_path=temp.name)
        

File: src/benchmarks/microbenchmarks/mbench_1.cpp
/*
*   @details:
*       1. Generate 64bit keys and 64bit values from an uniform random distribution to be inserted, removed or queried.
*       2. Evaluation of Insertion: from 0% -> filling_factor of the total capacity.
*       3. Evaluation of Queries: Positive and Negative queries (queries_factor) (success_factor).
*       4. Evaluation of Deletion: Remove a selections of keys until the usage reaches ~50%.
*
*   @author: Cristian Sandu <cristian.sandu@tum.de>
*/

#include "utils.hpp"
#include "hash.hpp"
#include <nlohmann/json.hpp>
#include <iostream>
#include <functional>
#include <vector>
#include <time.h>
#include <fstream>
#include <thread>

void*       libhandle = nullptr;
void*       generic_structure = nullptr;

void*       (*ds_init)(uint64_t);
void        (*ds_thread_init)(void*);
int         (*ds_insert)(void*, uint64_t, uint64_t);
int         (*ds_remove)(void*, uint64_t);
int         (*ds_read)(void*, uint64_t);
int         (*ds_read_range)(void*, uint64_t, uint64_t);
uint64_t    (*ds_get_size)(void*);

/* hashing function */
uint64_t            (*hash_fn)(uint64_t);

/* parameters */
std::vector<uint64_t> used_keys;
std::mutex            g_mutex;
Barrier               tbarrier;

/* data extracted from the program */
LogFile logfilePerformance("");
LogFile logfileMemory("");

uint64_t next_value() {
    static uint64_t current_value = 0;
    g_mutex.lock();
    int result = current_value++;
    g_mutex.unlock();
    return result;
}

void dataset_performfill(const size_t thread_id, void* ds,
                         const double filling_factor, const uint64_t maximum_capacity) {
    const uint64_t capacity = filling_factor * maximum_capacity;
    uint64_t status = 0;
    std::chrono::time_point<std::chrono::system_clock> start_time, end_time;

    for (uint64_t i=0; i<capacity; i++) {
        const uint64_t key_num  = next_value();
        const uint64_t key      = hash_fn(key_num);
        const uint64_t value    = hash_fn(next_value());

        MEASURE_TIME(ds_insert(ds, key, value),  start_time, end_time, status);
        logfilePerformance.add_log("dataset_performfill", thread_id, key_num, key, value, status, start_time, end_time);

        g_mutex.lock();
        uint64_t memory_usage = ds_get_size(ds);
        logfileMemory.add_log_mem("dataset_performfill", thread_id, key_num, memory_usage);
        used_keys.push_back(key_num);
        g_mutex.unlock();
    }
}

void dataset_performqueries(const size_t thread_id, void* ds,
                         const double query_factor, const double success_factor,
                         const uint64_t maximum_capacity) {
    const uint64_t capacity = query_factor * maximum_capacity,
                   success_capacity = success_factor * capacity;
    uint64_t status = 0;
    std::chrono::time_point<std::chrono::system_clock> start_time, end_time;    
    for (uint64_t i=0; i<capacity; i++) {
        uint64_t key_num = 0;
        if (success_capacity > 0 && i % (capacity / success_capacity) == 0) {
            g_mutex.lock();
            key_num   = used_keys[rand() % used_keys.size()];
            g_mutex.unlock();
        } else {
            key_num  = next_value();
        }

        const uint64_t key      = hash_fn(key_num);
        MEASURE_TIME(ds_read(ds, key),  start_time, end_time, status);
        logfilePerformance.add_log("dataset_performqueries", thread_id, key_num, key, 0, status, start_time, end_time);
    }
}

void dataset_performdeletion(const size_t thread_id, void* ds,
                         const double deletion_factor, const double success_factor,
                         const uint64_t maximum_capacity) {
    const uint64_t capacity = deletion_factor * maximum_capacity,
                   success_capacity = success_factor * capacity;
    uint64_t status = 0, success_count = 0;
    std::chrono::time_point<std::chrono::system_clock> start_time, end_time;    
    for (uint64_t i=0; i<capacity; i++) {
        uint64_t key_num = 0;
        if (success_capacity > 0 && i % (capacity / success_capacity) == 0) {
            g_mutex.lock();
            uint64_t index_random = rand() % used_keys.size();
            while(used_keys[index_random] == -1) { index_random = rand() % used_keys.size(); }

            key_num   = used_keys[index_random];
            used_keys[index_random] = -1;
            g_mutex.unlock();
        } else {
            key_num  = next_value();
        }
        const uint64_t key      = hash_fn(key_num);

        MEASURE_TIME(ds_remove(ds, key),  start_time, end_time, status);
        logfilePerformance.add_log("dataset_performdeletion", thread_id, key_num, key, 0, status, start_time, end_time);
    }
}

void benchmark_threads(const uint64_t threadid, nlohmann::json config_data, const uint64_t num_threads,
                       void *ds, const uint64_t capacity) {
    #ifdef DIFF_CPU_EXEC
        /* each thread will be executed on a different CPU */
        pthread_t self = pthread_self();
        cpu_set_t cpuset;
        CPU_ZERO(&cpuset);
        CPU_SET(threadid, &cpuset);
    #endif
    ds_thread_init(ds);

    const double   filling_factor  = config_data["performfill"]["filling_factor"];
    const uint64_t thread_capacity = capacity/num_threads;
    dataset_performfill(threadid, ds, filling_factor, thread_capacity);
    tbarrier.wait();

    const double  query_factor   = config_data["performquery"]["query_factor"];
    double  success_factor = config_data["performquery"]["success_factor"];
    dataset_performqueries(threadid, ds, query_factor, success_factor, thread_capacity);
    tbarrier.wait();

    const double  deletion_factor = config_data["performdeletion"]["deletion_factor"];
    success_factor  = config_data["performdeletion"]["success_factor"];
    dataset_performdeletion(threadid, ds, deletion_factor, success_factor, thread_capacity);
}

int main(int argc, char *argv[]) {
    if (argc != 4) {
        std::cerr << "[error] format: <path to so> <path to config file> <log file name>" << std::endl;
        exit(1);
    }
    srand(time(NULL));      /* prepare the randomness */

    std::string libpath   = argv[1];

    std::ifstream f(argv[2]);
    nlohmann::json config_data = nlohmann::json::parse(f);
    logfilePerformance.set_name(std::string(argv[3]) + "_performance.out");
    logfileMemory.set_name(std::string(argv[3]) + "_memory.out");

    const uint64_t num_threads = config_data["threadnum"];
    if (std::thread::hardware_concurrency() < num_threads) {
        std::cerr << "[error] the CPU has only " << std::thread::hardware_concurrency()
                << " while the threads num. is " << num_threads << std::endl;
        exit(1);
    }

    if (initialize(libpath) != 0) {
        std::cerr << "[error] failed: " << std::endl;
        exit(1);
    }

    if (config_data["distribution"] == "zipfian") {
        hash_fn = hash_zipfian;
    } else if (config_data["distribution"] == "uniform") {
        hash_fn = hash_uniform;
    } else {
        throw std::runtime_error("unknown distribution type");
    }

    /* thread barrier initialization */
    tbarrier.set_numthreads(num_threads);

    /* all functions have been loaded at this point */
    const uint64_t capacity  = config_data["capacity"];
    void* ds = ds_init(capacity);
    
    const uint64_t thread_capacity = capacity/num_threads;
    EXECUTE_PARALLEL(num_threads, benchmark_threads, config_data, num_threads, ds, capacity);

    logfilePerformance.save_logfile();
    logfileMemory.save_logfile();
    return 0;
}